server/src/main/java/org/elasticsearch/search/aggregations/bucket/filter/QueryToFilterAdapter.java

```diff
@@ -0,0 +1,343 @@
+            return Lucene.asSequentialAccessBits(ctx.reader().maxDoc(), weight().scorerSupplier(ctx))::get;
+        }
+
+        @Override
+        long count(LeafReaderContext ctx, FiltersAggregator.Counter counter, Bits live) throws IOException {
+            BulkScorer scorer = bulkScorer(ctx, () -> {});
+            if (scorer == null) {
+                // No hits in this segment.
+                return 0;
+            }
+            scorer.score(counter, live);
+            return counter.readAndReset(ctx);
+        }
+
+        @Override
+        long estimateCountCost(LeafReaderContext ctx, CheckedSupplier<Boolean, IOException> canUseMetadata) throws IOException {
+            BulkScorer scorer = bulkScorer(ctx, () -> scorersPreparedWhileEstimatingCost++);
+            if (scorer == null) {
+                // There aren't any matches for this filter in this leaf
+                return 0;
+            }
+            return scorer.cost();
+        }
+
+        private BulkScorer bulkScorer(LeafReaderContext ctx, Runnable onPrepare) throws IOException {
+            if (bulkScorers == null) {
+                bulkScorers = new BulkScorer[searcher().getIndexReader().leaves().size()];
+            }
+            if (bulkScorers[ctx.ord] == null) {
+                onPrepare.run();
+                return bulkScorers[ctx.ord] = weight().bulkScorer(ctx);
+            }
+            return bulkScorers[ctx.ord];
+        }
+
+        private Weight weight() throws IOException {
+            if (weight == null) {
+                weight = searcher().createWeight(query, ScoreMode.COMPLETE_NO_SCORES, 1.0f);
+            }
+            return weight;
+        }
+
+        @Override
+        void collectDebugInfo(BiConsumer<String, Object> add) {
+            super.collectDebugInfo(add);
+            add.accept("query", query.toString());
+            add.accept("scorers_prepared_while_estimating_cost", scorersPreparedWhileEstimatingCost);
+        }
+    }
+
+    private static class DefaultQueryToFilterAdapter extends CommonQueryToFilterAdapter<Query> {
+        private DefaultQueryToFilterAdapter(IndexSearcher searcher, String key, Query query) {
+            super(searcher, key, query);
+        }
+
+        @Override
+        void collectDebugInfo(BiConsumer<String, Object> add) {
+            super.collectDebugInfo(add);
+            add.accept("type", "query");
+        }
+    }
+
+    private static class TermFilter extends CommonQueryToFilterAdapter<TermQuery> {
+        private int resultsFromMetadata;
+
+        private TermFilter(IndexSearcher searcher, String key, TermQuery query) {
+            super(searcher, key, query);
+        }
+
+        @Override
+        long count(LeafReaderContext ctx, FiltersAggregator.Counter counter, Bits live) throws IOException {
+            if (countCanUseMetadata(counter, live)) {
+                resultsFromMetadata++;
+                return ctx.reader().docFreq(query().getTerm());
```

Comment:
This right here is the magic.

We certainly could have done with without adding this whole abstraction. But the abstraction really helps me reason about what we want from queries and having a subclass per query that has a fancy optimization helps me to identify which fancy optimizations we have. And it helps us track how they are being applied when you profile it.